Week5 [STEP 1] 이준호

junho15 opened the pull request:
@siwonkim0 
마지막 PR 이네요. 5주가 지나고보니 짧습니다. 그동안 감사했습니다.

## 고민했던 점
Person 클래스, Level 열거형, Talent, BadPersonality 프로토콜을 추가했습니다.

Person을 상속하고 Talent를 준수하는 TalentedPerson 클래스와 Person을 상속하고 Talent, BadPersonality를 준수하는 TalentedPersonWithBadPersonality 클래스를 만들었습니다.

프로토콜 요구사항으로 만든 프로퍼티(singing, dancing 등)는 private 키워드를 추가할 수가 없네요. 우선은 상수로 해서 수정을 못하게 했습니다.

AuditionManager 구조체를 만들어 cast(), announcePassedApplicants() 메서드를 추가했습니다.
cast 메서드에서는 if let 과 as? 를 사용해서 다운캐스팅을 시도했습니다. 
isGoodPerson 메서드로 합격 여부를 판단해서 passedApplicantsList에 추가하도록 했습니다. 

main에서는 지원자 인스턴스를 만들고 AuditionManager를 생성해서 cast(), announcePassedApplicants 메서드 호출했습니다.

junho15 committed:
feat: Person, TalentedPerson, TalentedPersonWithBadPersonality 클래스 추가 외

Person, TalentedPerson, TalentedPersonWithBadPersonality 클래스 추가
Level 열거형 추가
Talent, BadPersonality 프로토콜 추가
AuditionManager 구조체 추가
지원자, AuditionManager 생성 및 cast(), announcePassedApplicants 메서드 호출

siwonkim0 commented on CodeStarterCamp_Week5/AuditionManager.swift:
```diff
@@ -0,0 +1,40 @@
+    }
+    
+    
+    mutating func cast() {
+        for applicant in totalApplicantsList {
+            if let talentedPersonWithBadPersonality = applicant as? TalentedPersonWithBadPersonality {
+                if talentedPersonWithBadPersonality.isGoodPerson() {
+                    passedApplicantsList.append(talentedPersonWithBadPersonality)
+                }
+            }
+            if let talentedPerson = applicant as? TalentedPerson {
+                if talentedPerson.isGoodPerson() {
+                    passedApplicantsList.append(talentedPerson)
+                }
+            }
+        }
+    }
```
cast() 메서드에 if 문이 4개나 있네요! 한 함수가 많은 기능을 하게 되면 가독성도 떨어지고, 재사용성도 떨어지게 되는데요.
각 if문을 따로 함수로 분리할 수 있을 것 같습니다.
(이렇게 분리된 함수의 경우에는 타입 내부에서만 불리기 때문에 private 키워드를 붙일 수 있겠죠?)

또한 for문의 경우 고차함수를 사용해서 변경해볼까요?

siwonkim0 commented on CodeStarterCamp_Week5/TalentedPerson.swift:
```diff
@@ -0,0 +1,30 @@
+        super.init(name: name, height: height)
+    }
+    
+    func isGoodPerson() -> Bool {
```
함수 네이밍이 함수가 하는 일과 다른 것 같아요. isGoodPerson은 성격이 좋은지 체크하는 기능을 가질 것 같은데 함수가 하는 일은 재능이 통과될만한지 체크하는거니 이에 맞는 네이밍을 해줄까요?

siwonkim0 commented on CodeStarterCamp_Week5/Talent.swift:
```diff
@@ -0,0 +1,14 @@
+
+import Foundation
+
+protocol Talent {
+    var singing: Level { get }
+    var dancing: Level { get }
+    var acting: Level { get }
+}
```
let 프로퍼티를 표현하려면 프로토콜에는 var { get }로 정의해야해죠! 좋습니다 👍

siwonkim0 commented on CodeStarterCamp_Week5/main.swift:
```diff
@@ -8,5 +8,17 @@
+                                             acting: .A, frequancyOfCursing: .A)
 
+var auditionManager = AuditionManager(applicansts: [yagom, noroo, summer, coda, odong])
+auditionManager.cast()
+auditionManager.announcePassedApplicants()
```
cast() 메서드의 역할을 분리해주셨다면, 캐스팅된 후보를 프린트하는announcePassedApplicants() 메서드도 cast() 메서드 안에서 호출해볼까요? 항상 캐스팅 된 후 결과를 프린트하기때문에 같이 묶어도 좋을 것 같습니다 :)

siwonkim0 reviewed (commented):
안녕하세요 준호님~ 
마지막 주차 마지막 PR이네요 ㅠㅠ 5주가 이렇게 빠르다니 아쉽습니다...

전체적으로 코드가 가독성있고 깔끔하기도 하고 프로토콜과 타입 캐스팅도 적절하게 사용해주셨어요 👍
사소한 부분인데 함수분리에 대한 부분 피드백 드렸습니다 :)
편하게 수정해주시고 답글 주세요 ㅎㅎ

감사합니다😁

siwonkim0 commented:
하나만 더 말씀드리자면, 현재 커밋 단위가 너무 큰 것 같습니다. 
규모가 작은 프로젝트라서 지금은 상관없을수도 있지만, 규모가 커지면 힘들어지기때문에
초반에 커밋을 작은 단위로 하는 습관형성을 하는 것이 중요하다고 생각해요

저도 초반에 공부할때 커밋 단위를 어떻게 쪼개야할지 몰라서 고민했던 경험이 있습니다.
저같은 경우에는 보통 한 기능을 완성한 후 커밋을 하는데요, 
선배 캠퍼에게 "날려도 금방 다시 복구할 수 있는 수준의 단위"로 커밋하라는 조언을 들었던 기억이 있네요 ㅎㅎ

지금 프로젝트를 예시로 들자면 저라면 타입을 하나 만들때마다 커밋을 했을 것 같습니다. 
그리고 타입을 다 만들고 나서 타입에 메서드를 추가할 일이 있다면 메서드 하나 만들고 커밋하는 식으로 할 것 같아요 :)

커밋 단위를 잘 쪼개면 나중에 코드를 다시 볼때도 수정 사항이 단위별로 잘 나눠져있어서 기억하기도 쉽답니다 ㅎㅎ

junho15 committed:
fix: isGoodPerson 메서드 이름  isPassedPerson으로 수정

junho15 committed:
fix: for문 고차함수 compactMap으로 변경 외

for문 고차함수 compactMap으로 변경
isPassedTalentedPersonWithBadPersonality, isPssedTalentedPerson 메서드 추가해서 분리

junho15 committed:
fix: cast 메서드 안에서 announcePassedApplicants 메서드 호출
